### DISCUSSION TITLE
Using IDE session endpoint requests outside of Aspire

### DISCUSSION THREAD

**Discussion by epsitec** [microsoft/aspire#2355] [Q&A] [answered]
We've been experimenting with Aspire and found that the `DEBUG_SESSION_PORT` and `DEBUG_SESSION_TOKEN` environment variables would be very helpful in our own code, in order to help debugging a multi-project solution. Specifically, we'd like to use the Visual Studio [IDE-execution API described here](https://github.com/dotnet/aspire/blob/main/docs/specs/IDE-execution.md), without having to include `Aspire.Hosting` .

Somehow, adding `<IsAspireHost>true</IsAspireHost>` does some magic behind the scenes, which makes Visual Studio provide both `DEBUG_SESSION_PORT` and `DEBUG_SESSION_TOKEN` environment variables. How can we trigger the same behavior, without setting `<IsAspireHost>` in the `*.csproj`?

**Comment by davidfowl** [comment]
How are you launching multiple projects in the first place?

cc @BillHiebert @karolz-ms

**Reply by epsitec** [reply]
We've been using two approaches up to now, none being really satisfying.

- Visual Studio's _Configure Startup Projects_ helps in solutions where projects can be launched independently one from another; we have to cheat if there are dependencies and hardwire the arguments. For our current scenario, this is not ideal, since we have to predefine the ports on which the various processes operate and can't use the dynamic port allocation used in production, and we have to add delays when starting some projects, replace `Process.Start()` with dummy code in our orchestrator, etc.
- Starting only the main project from Visual Studio, then attaching afterwards to the other processes, either manually through _Debug_ > _Attach to process_, or by adding calls to `Debugger.Break()` in every `Program.cs`. But this is not really satisfying...

After discovering Aspire's use of the IDE-Execution API (and `DEBUG_SESSION_PORT` + `DEBUG_SESSION_TOKEN`) we've used it to launch child processes by asking the IDE to start their associated `.csproj` file. However, this only works when Aspire's magic kicks in (`<IsAspireHost>true</IsAspireHost>`). We don't want to include Aspire itself into our solution (please assume we have some valid reason for this ;-) or else I wouldn't have to ask). Removing the project reference on Aspire seems to have unexpected side effects, such as the fact that referenced assemblies no longer get properly referenced (missing references). I suppose this has something to do with how the Aspire SDK rewires the build process to generate the dependencies and manifests.

Basically, I'd be very happy to be able to have Visual Studio start my main project without Aspire, while still making its IDE-Execution API available to it. Is this possible? What is the trick which makes Visual Studio activate this feature?

**Reply by davidfowl** [reply, marked as answer]
So you’re building your own orchestrator/launcher and you want to hook into the debugger the same way aspire does?

You probably need this project capability to begin with. I think this is what makes VS launch with the debug session token and port. Then you have to implement the protocol.

The project capability is here:

https://github.com/dotnet/aspire/blob/ceae04d46c5e0aa44274776a8463210a1aa31f9b/src/Aspire.Hosting.Sdk/SDK/Sdk.targets#L12

The protocol is here

https://github.com/dotnet/aspire/blob/main/docs/specs/IDE-execution.md

Hope that helps

**Reply by epsitec** [reply]
Switching from:
```csproj
<ProperyGroup>
  <NoWarn>ASPIRE002</NoWarn>
  <IsAspireHost>ture</IsAspireHost>
</PropertyGroup>
```
to
```
<ItemGroup>
  <ProjectCapability Include="AspireOrchestration" />
</ItemGroup>
```
did indeed solve my issues.

- Starting the main project indeed injects the `DEBUG_SESSION_...` environment variables.
- Project references through `<ProjectReference Include="..."/>` work as expected.

Thank you very much!

**Reply by davidfowl** [reply]
No problem!

**Comment by idg10** [comment]
This has always seemed like a missing feature of Visual Studio. Clearly other people also feel this way because this is a thing: https://marketplace.visualstudio.com/items?itemName=vsdbgplat.MicrosoftChildProcessDebuggingPowerTool2022 (and its [predecessor](https://marketplace.visualstudio.com/items?itemName=vsdbgplat.MicrosoftChildProcessDebuggingPowerTool) has existed since 2014), but it has always been a bit hacky because it requires you to enable mix mode debugging.

When I realised Aspire was somehow managing to attach the debugger to all of the processes it launched, I wondered if that meant the IDE was finally offering this as a built-in service. But from this it seems like the Aspire team had to build their own workaround.

The fact that we can enable this with `<ProjectCapability Include="AspireOrchestration" />` and not need to be building an Aspire project to use this is awesome, but it's something I'd be a bit concerned about depending on. The [list of project capabilities](https://github.com/microsoft/VSProjectSystem/blob/master/doc/overview/project_capabilities.md) does not include `AspireOrchestration`. Is the ability to use this outside of an Aspire solution a documented feature? Or just a really cool hack that might stop working at some point?
